MainPage 상품리스트 endpoint 구현

jay95ko opened the pull request:
>wecode PR message template 및 체크 리스트 입니다. 
아래 사항들을 전부 작성/체크 하시고 PR 해주세요!

## 수정 사항 간략한 한줄 요약
(여기에 수정 사항에 대한 간략한 한줄 요약/제목 작성해 주세요.)
- 모든 요청에 대해 DB hit을 최소화 하기 위해 ORM수정
- Category에 대한 정보 또한 API로 요청하여 받기 위해 해당 endpoint 추가
		 
## 수정 사항들 자세한 내용
(여기에 수정 사항에 대한 자세한 내용을 작성해 주세요.)
- Query를 확인 하는 logger 생성
- 현재 까지 작성했던 endpoint에 대하여 장고ORM 최적화(select_related, predetch_related)
- category별 image에 대한 endpoint 생성

## 기타 질문 및 특이 사항
(궁금한 사항들, 하면서 느낀 점들 공유해주실 것이 있으면 자유롭게 작성해 주세요.)

## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
- [x] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [x] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [x] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [x] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.

jay95ko committed:
Add
Add:
views.py -> MainPageView
Make recommend_product_list

jay95ko committed:
Add
Add: products.views.py ProductMainPageView
Make recommend_product_list

jay95ko committed:
Add: sale_product

jay95ko committed:
Add: popular_product_list

jay95ko committed:
Add: new_product_list

jay95ko committed:
Add, Modify
Make MainPage_product_list
Add: MakeList function
	parameter : QuerySet
	OutPut : list
Modify: Delete code becuz use function

lordmyshepherd commented on products/urls.py:
```diff
@@ -0,0 +1,7 @@
+from .views import ProductsMainPageView
+
+urlpatterns = [
+    path("/main", ProductsMainPageView.as_view()),
```
"MainPage에 필요한 데이터를 반환해야 된다."라는 생각 때문에 MainPageView 라는 class명을 사용하신 것 같습니다.
API는 페이지를 기준으로 만드시면 안됩니다. API는 자원(data)를 기준으로 작성해주셔야 합니다. 즉, 엔드포인트는 페이지에 의존성을 갖도록 만들면 안되고 메인 페이지 뿐만 아니라, 다른 페이지에서도 상품 데이터가 필요하면 아래와 같이 요청해서 받아갈 수 있도록 만들어주셔야 합니다.

```python
GET :8000/products -> 상품 리스트를 받아오는 구나!!
GET :8000/product -> 상품을 받아오는 구나!!
```
Restful API 세션을 듣고 수정해주세요~

lordmyshepherd commented on products/views.py:
```diff
@@ -1,3 +1,85 @@
+
+from .models import Product
+
+class ProductsMainPageView(View):
```
class명도 `ProductsView` 또는 `ProductListView`로 수정해주세요~

lordmyshepherd commented on products/views.py:
```diff
@@ -1,3 +1,85 @@
+from .models import Product
+
+class ProductsMainPageView(View):
+    def get(self, request):
```
해당 엔드포인트의 목적이 `recommend_product_list`, `sale_product_list`, `popular_product_list`, `new_product_list`를 반환해준다고 하면, 구조를 수정해야될 것 같습니다.

아마도, 메인 페이지에서 추천상품, 세일상품, 인기상품, 신상품을 한번에 보여주기 위해서 한번에 모든 데이터를 전부 반환해주도록 만드신 것 같습니다. 이렇게 하시면,

1. `추천상품`만 데이터가 필요한 경우에도 전부 받아와야 하는 문제가 발생할 수 있습니다.
2. 필터링 기능을 구현할 때 비슷하거나 같은 코드를 중복으로 또 작성해서 새로운 엔드포인트를 만들어야 합니다.

결론적으로 `query paramet( ../products?filter=new)`를 사용해서 해당 키워드로 요청이 들어왔을 때, `filtering`해서 반환할 수 있도록 구현해주시고, 프론트에서 `?filter=new`와 `?filter=recommand` 같이 총 4번 요청할 수 있도록 만들어주세요~

lordmyshepherd replied:
위에서 말씀드린 피드백을 기반으로 아래 구조를 전반적으로 수정해주세요.
for문의 사용을 최소한으로 혹은 사용하지 않는 방향으로 고민해보세요

jay95ko replied:
네 이해했습니다!! 수정해여 다시 PR하겠습니다

jay95ko committed:
Add: use query_parameter

jay95ko committed:
Product API with Restful
Add:
urls.py, views.py
상품개별, 상품리스트, 삼품디테일 페이지에서 필요한 요소 각각 endpoint 구현

jay95ko commented:
이전의 리뷰에서 말씀주신 내용과 더불에 Restful API의 형식에 맞춰 구현하였습니다.

jay95ko committed:
Modify ProductListView
parameter중 카테고리가 여러개 들어 왔을 때 검색이 가능하도록 수정

jay95ko committed:
delete ForLoop
제거 가능한 for loop삭제

lordmyshepherd commented on products/views.py:
```diff
@@ -1,3 +1,157 @@
+class ImageListView(View):
+    def get(self, request, product_id):
+        try:
+
```
line 38 빈 줄 지워주세요~

lordmyshepherd commented on products/views.py:
```diff
@@ -1,3 +1,157 @@
+    def get(self, request, product_id):
+        try:
+
+            images = list(Product.objects.get(id=product_id).images.all().values("image_url"))
```
ImageView는 가독성과 데이터베이스 요청 쿼리 수를 줄이기 위해서 Image 객체부터 접근해서 가져올 수 있도록 수정해주세요~
```suggestion
            images = Image.objects...

```

lordmyshepherd replied:
참고) 왜 작성하신 방법이 database 요청 수가 더 많은지 고민해보시면 좋을 것 같아요~

lordmyshepherd commented on products/views.py:
```diff
@@ -1,3 +1,157 @@
+class BreweryView(View):
+    def get(self, request, product_id):
+        try:
+            product = Product.objects.get(id=product_id)
+
+            breweries = []
+            breweries.append(
+                {
+                    "brewery_name" : product.brewery.name,
+                    "brewery_image": product.brewery.img_url,
+                    }
+                )
+            return JsonResponse({"Result": breweries}, status=200)
```
여기도 위에서 드린 피드백과 동일합니다~

lordmyshepherd commented on products/views.py:
```diff
@@ -1,3 +1,157 @@
+            return JsonResponse({"Result": "PRODUCT_DOES_NOT_EXIST"}, status=400)
+
+class FlavorListView(View):
+    def get(self, request, product_id):
```
여기도 위에서 드린 피드백과 동일합니다~

lordmyshepherd commented on products/views.py:
```diff
@@ -1,3 +1,157 @@
+class SidedishListView(View):
+    def get(self, request, product_id):
+        try:
+            product   = Product.objects.get(id=product_id)
```
여기도 위에서 드린 피드백과 동일합니다~

lordmyshepherd commented on products/views.py:
```diff
@@ -1,3 +1,157 @@
+            return JsonResponse({"Result": "PRODUCT_DOES_NOT_EXIST"}, status=400)
+
+class DetailView(View):
+    def get(self, request, product_id):
```
여기도 위에서 드린 피드백과 동일합니다~

lordmyshepherd commented on products/views.py:
```diff
@@ -1,3 +1,157 @@
+        except Product.DoesNotExist:
+            return JsonResponse({"Result": "PRODUCT_DOES_NOT_EXIST"}, status=400)
+
+class ProductView(View):
```
ProductVIew PR 분리해주세요!

lordmyshepherd commented on products/views.py:
```diff
@@ -1,3 +1,157 @@
+        except Product.DoesNotExist:
+            return JsonResponse({"Result": "PRODUCT_DOES_NOT_EXIST"}, status=400)
+
+class ProductListView(View):
```
여기도 PR 분리해주세요~
PR은 반드시 하나의 기능만 올려주셔야해요! 그렇지 않으면 아래와 같은 문제점들이 있습니다.

1. 팀원들의 리뷰를 받을 때, 너무 많은 코드와 기능을 파악하고 리뷰를 해야된다. -> 당연히 제대로된 피드백을 받기 어려워진다.
2. 6개의 기능 중에 3개는 merge되도 되는 기능인데도 나머지 3개의 기능의 수정사항 때문에 계속해서 기능 완료 시점이 지연된다.

lordmyshepherd replied:
아직 피드백 반영이 안된 것 같습니다!

jay95ko replied:
현재 PR은 종료하고 다시 새로 PR올리겠습니다!!

jay95ko committed:
Modify
수정요청사항 반영
기능별 PR분리
데이더베이스 접근 요청 최소화

jay95ko commented:
말씀 주신 사항 반영하였으며, PR분리하겠습니다!

jay95ko committed:
Add: Product endpoint

jay95ko committed:
Add: Products endpoint

jay95ko commented:
해당 PR이 Merge 되지 않아 분리가 안되어서 commit단위로 분리하였습니다.

참고로 남겨주신 고민사항은 오늘 내려오시면 여쭤보겠습니다!!

jay95ko committed:
Modify: for convention

jay95ko committed:
Add: products add filter

jay95ko committed:
commit for merge confilic

jay95ko committed:
ORM 최적화 및 추가기능사항 구현
1. 모든 요청에 대해 DB hit을 최소화 하기 위해 ORM수정(select_related, prefetch_related)
1-1. 해당 DB요청 Query보기 위해 loggger추가
1-2. logger확인하며, Query 현저히 줄어드는 것 확인
2. Category에 대한 정보 또한 API로 요청하여 받기 위해 해당 endpoint 추가

jay95ko commented:
병민님 말씀주셨던 사항 관련하여, 공부하고 최대한 DB hti이 적은 방향으로 생각하여 수정해 보았습니다.
확인 및 리뷰 부탁드립니다!

jay95ko committed:
Modify: for -> list comprehention

jay95ko commented:
for문 보다는 list comprehention으로 표현 할 수 있는 부분은 수정 하였습니다.

jay95ko committed:
Modify: FE에서 데이터를 가져감에 있어 효율적인 방법으로 변경

jay95ko committed:
Modify: QuerySet을 list화 시키기 위해 []썼으나 변경이 안되어 수정
